install/0000_90_machine-config-operator_01_prometheus-rules.yaml

```diff
@@ -93,13 +93,63 @@ spec:
-    - name: master-nodes-high-memory-usage
+    - name: high-overall-control-plane-memory
       rules:
-        - alert: MasterNodesHighMemoryUsage
+        - alert: HighOverallControlPlaneMemory
           expr: |
-            ((sum(node_memory_MemTotal_bytes AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )) - sum(node_memory_MemFree_bytes + node_memory_Buffers_bytes + node_memory_Cached_bytes AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" ))) / sum(node_memory_MemTotal_bytes AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )) * 100) > 90
-          for: 15m
+            (
+              1
+              -
+              sum (
+                node_memory_MemFree_bytes
+                + node_memory_Buffers_bytes
+                + node_memory_Cached_bytes
+                AND on (instance)
+                label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+              ) / sum (
+                node_memory_MemTotal_bytes
+                AND on (instance)
+                label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+              )
+            ) * 100 > 60
+          for: 1h
           labels:
             severity: warning
           annotations:
-            message: "Memory usage of {{ $value | humanize }} on {{ $labels.node }} exceeds 90%. Master nodes starved of memory could result in degraded performance of the control plane."
+            summary: >-
+              Memory utilization across all control plane nodes is high, and could impact responsiveness and stability.
+            description: >-
+              Given three control plane nodes, the overall memory utilization may only be about 2/3 of all available capacity.
+              This is because if a single control plane node fails, the kube-apiserver and etcd my be slow to respond.
+              To fix this, increase memory of the control plane nodes.
+    - name: extremely-high-individual-control-plane-memory
+      rules:
+        - alert: ExtremelyHighIndividualControlPlaneMemory
+          expr: |
+            (
+              1
+              -
+              sum by (instance) (
+                node_memory_MemFree_bytes
+                + node_memory_Buffers_bytes
+                + node_memory_Cached_bytes
+                AND on (instance)
+                label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+              ) / sum by (instance) (
+                node_memory_MemTotal_bytes
+                AND on (instance)
+                label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
+              )
+            ) * 100 > 90
+          for: 45m
```

Comment:
45m diverges from the "Timeline: ..." recommendations in [the policy doc][1].  I don't have a problem with it as a value, but if we have any details about how we came up with this number, we should put them in the commit message where future devs who are tempted to adjust the number have a chance of finding them.  The commit message as a whole could probably use some filling out.  Something like:

```
The docs recommend limiting control plane memory consumption to <=60% [1].  This commit adds HighOverallControlPlaneMemory to give warning-level alerting when the control plane as a whole passes that level.  Scaling the control plane instances currently requires draining and rebooting nodes, so we don't want to wait for higher resource consumption before scaling up.  60m is the default 'for' for warning alerts [2], and we have no reason to diverge from that default here.

Also replace the old, full-control-plane MasterNodesHighMemoryUsage with ExtremelyHighIndividualControlPlaneMemory at the >90% level, and raise severity to critical.  The new name aligns with the established CPU-alerting pattern.  The new severity is compatible with the alerting policy [3], because manual pod-rescheduling or node-scale-ups would be required to preserve etcd throughput and control-plane functionality in such a low-free/cache/buffer situation.  45m diverges from the default 'for: 5m' for critical alerts [3], because... REASONS.

[1]: https://docs.openshift.com/container-platform/4.10/scalability_and_performance/recommended-host-practices.html#master-node-sizing_recommended-host-practices
[2]: https://github.com/openshift/enhancements/blob/34ba81cd125d788062172ca4c630bd228d1609bf/enhancements/monitoring/alerting-consistency.md#warning-alerts
[3]: https://github.com/openshift/enhancements/blob/34ba81cd125d788062172ca4c630bd228d1609bf/enhancements/monitoring/alerting-consistency.md#critical-alerts
```

or something like that.  I'm not personally committed to `critical` `for: 45m` for the >90% single-node level.  It feels like we're trying to dance on the edge between "more important/time-sensitive than `warning`" and "we don't want to wake admins up for a 10m spike".  But whatever our chosen code ends up being, the commit message should try to explain those choices.

[1]: https://github.com/openshift/enhancements/blob/34ba81cd125d788062172ca4c630bd228d1609bf/enhancements/monitoring/alerting-consistency.md

Reply:
FWIW, the example in that doc also doesn't match its own timeline.

Reply:
yeah, I don't think there's a strong need to match the recommendations.  Just that folks should explain what they chose, whether that's "this is the policy default" or "this is not the policy default, because ..." or whatever.

Reply:
updated

Reply:
Waiting 45 minutes before notifying the administrators when the control plane is at 90% CPU sounds like a lot. At that point, we can expect the apiserver to be very slow to respond which may have heavy consequences for the overall cluster and its availability.

Reply:
Besides during cluster startup, I don't think any normal behavior should be causing the control plane to be at 90% CPU usage. If that's the case, it is a good indicator that needs to increase the number of resources allocated to their cluster because if they ever experience a slight increase in traffic, the cluster won't be able to support it at all.

To emphasize more on that, you already added an alert on 60% usage which is meant to indicate to cluster admins that their resource allocation isn't ideal and they need to increase it. If 60% is the baseline at which we tell customers that their configuration will not scale, then 90% is a critical state that needs to be handled asap to prevent any kind of degradation.

Reply:
@dgrisonnet ok, May I ask what is the duration you recommend so that I make a new PR ?

Reply:
@dofinn @mmazur Hello ✋🏽 above @dgrisonnet  suggested duration for the critical alert .. Would you confirm ?

I will raise a fix once we agree on a duration :)

Reply:
For some reason, I thought this was alerting on CPU usage and superseding https://github.com/openshift/machine-config-operator/pull/3052. That said, my previous comments still apply. High memory usage and CPU usage have the same severity.